### ISSUE TITLE
Report all existing times and locations or only those used in mentions?

### ISSUE THREAD

**Issue by kwalcock** [clulab/eidos#635]
The eidos code has been nicely transitioned to use finders to find the times and locations.  These are then recorded in mentions.  Temporarily we've been adding a second copy of the times and locations to an EidosDocument to preserve compatability with previous versions.  We'd like to remove the EidosDocument and probably the second copy of the times and locations that it stores.

The problem with doing this is that this second copy is far more complete than the copy in the collection of Mentions because many of the original mentions get thrown away during extraction because they are not used in relationships or whatever else we're interested in (AFAICT).  Until now we've been including the complete list of times and locations divvied up by sentence in the JSONLD.  The first document I looked at had 19 times, but only 3 of them were used in the end and appear in the JSONLD when we don't rely on the the second copy.  Another place not all would appear is the WebApp.  (For that it might be possible to run the TimeNormFinder or GeoNormFinder individually to get the entire list.)

Is this the desired behavior: show only those times and locations involved in a reported mention?  Should I find somewhere else to store the complete list instead?

**Comment by MihaiSurdeanu** [comment]
This is not too bad of an outcome... But why don't we create EidosMentions from all the times and locations picked by the finders? Those should be preserved in the output without major changes, no?

**Comment by bethard** [comment]
The timenorm and geonorm finders already create Mentions for everything. Does Eidos lose those somehow?

**Comment by kwalcock** [comment]
It apparently has to do with the `cagRelevantOnly` setting, which defaults to true.  So there we are intentionally throwing things away.  Previously the times and locations were unaffected by this setting and all were included with the JSONLD output.  I guess that's OK.  It's a little strange to include them in the sentence-level output then.  With it false, I'm still losing one of the 19 and trying to track it down.  I wonder if the webapp still needs any separate output for times or locations since they will be included with the mentions now.

**Comment by bethard** [comment]
Yeah, I was wondering the same thing. My hope was that we wouldn't need any separate stuff for them anymore, and we could just output them the same way due with other mentions.

**Comment by MihaiSurdeanu** [comment]
I see. I think we should keep the semantics of `cagRelevantOnly`. But are you saying that you are losing one of the times/locations that is attached to a causal relation? If so, that is a bug.

Also, can users of the web app adjust the value of `cagRelevantOnly` in the app? I don't recall if it's possible.

**Comment by kwalcock** [comment]
The WebApp does have a checkbox for cagRelevantOnly, so maybe that takes care of it.

There might be a bug.  A time is lost when cagRelevantOnly is false and they shouldn't be filtered out.  I'll be tracking it down.

**Comment by kwalcock** [comment]
EidosActions.keepMostCompleteEvents is able to remove Mentions that have Time attachments so that not all of the originally discovered times will be available after event extraction, even if cagRelevantOnly is false.  I don't know that this is terrible, but it is changing the previous behavior.

**Comment by bethard** [comment]
Maybe when it is filtering the most complete events it should copy over any attachments from events it discards?

**Comment by kwalcock** [comment]
That's an idea.  Move over to where, though?  I think @BeckySharp would need to weigh in on it.

**Comment by BeckySharp** [comment]
sorry to take so long to weigh in guys! 
The filtering for complete events is separate from the filtering for CAG relevant.  In the former, we discard (hopefully!) duplicates, and I think the "most complete" (filter survivor) should have all the attachments from any of the subsumed guys that were discarded.
In the latter, we toss away some concepts that were never part of a causal event, and in that case I think we either want to also discard the times in that concept, or else we should be keeping the concept -- right?

**Comment by kwalcock** [comment]
I'm not sure this is always happening: ` the "most complete" (filter survivor) should have all the attachments from any of the subsumed guys that were discarded.`  Is it important enough that I try to track down specific examples?

**Comment by BeckySharp** [comment]
i dont know.... maybe a better use of time would be to re-write the monstrosity that is that function(s) from scratch such that it's true....?

**Comment by kwalcock** [comment]
Then a specific failing example would be required, then a re-write, and then a green bar.  How about I find an example and then file an issue for documentation and consideration?

**Comment by BeckySharp** [comment]
Sure if that’s not too difficult/burdensome/annoying 
If at some point you decide it is let me know and we can regroup?  Remember, we want you to *enjoy* working on WM!

**Comment by kwalcock** [comment]
Starting with text from the attached file, some 29 TimEx of the 913 expressions that are initially found do not survive until the end.  This almost always happens when the `contextSentenceIndexes` are not matched up with word boundaries.  The list below tries to demonstrate the problems.  The lines with text/startOffset/endOffset come from the parsed/tokenized words in the sentence.  There are also interesting cases where one TimEx found by TimeNormFinder is inside another.  In these cases, only one of the two survives.

```
"TimEx(Interval(10714, 10721),List(),January)"
"text" : "January-December",
"startOffset" : 10714,
"endOffset" : 10730,

// One TimEx is inside another
"TimEx(Interval(14233, 14264),List(),30 November to 18 December 2015)"
"TimEx(Interval(14245, 14250),List(),to 18)" <- missing then

"TimEx(Interval(15938, 15941),List(),day)"
"text" : "12-day",
"startOffset" : 15935,
"endOffset" : 15941,

"TimEx(Interval(20638, 20646),List(),December)"
"text" : "January-December",
"startOffset" : 20630,
"endOffset" : 20646,

"TimEx(Interval(21917, 21918),List(),r)"
"text" : "year",
"startOffset" : 21914,
"endOffset" : 21918,

// One TimEx is inside another
"TimEx(Interval(25890, 25920),List(),between July and December 2015)"
"TimEx(Interval(25898, 25902),List(),July)" <- missing then

"TimEx(Interval(31508, 31510),List(TimeStep(0006-01-01T00:00,0007-01-01T00:00)),16)"
"text" : "2015/16",
"startOffset" : 31503,
"endOffset" : 31510,

"TimEx(Interval(32016, 32020),List(TimeStep(2015-01-01T00:00,2016-01-01T00:00)),2015)"
"text" : "2015/16",
"startOffset" : 32016,
"endOffset" : 32023,

"TimEx(Interval(32673, 32675),List(TimeStep(0016-01-01T00:00,0017-01-01T00:00)),16)"
"text" : "2015/16",
"startOffset" : 32668,
"endOffset" : 32675,

// One TimEx is inside another
"TimEx(Interval(33882, 33889),List(TimeStep(0216-01-01T00:00,0217-01-01T00:00)),2015/16)" <- missing
"TimEx(Interval(33885, 33886),List(),5)"


"TimEx(Interval(33942, 33944),List(TimeStep(0015-01-01T00:00,0016-01-01T00:00)),15)"
"text" : "2014/15",
"startOffset" : 33937,
"endOffset" : 33944,

"TimEx(Interval(38117, 38119),List(),ar)"
"TimEx(Interval(38123, 38127),List(),year)"
"text" : "year-on-year",
"startOffset" : 38115,
"endOffset" : 38127,

"TimEx(Interval(52057, 52065),List(),December)"
"text" : "November-December",
"startOffset" : 52048,
"endOffset" : 52065,

// One TimEx is inside another
"TimEx(Interval(67147, 67153),List(),August)" <- missing
"TimEx(Interval(67142, 67153),List(),July/August)"

// One TimEx is inside another
"TimEx(Interval(68211, 68226),List(),timely in March)"
"TimEx(Interval(68215, 68217),List(),ly)" <- missing

// One TimEx is inside another
"TimEx(Interval(68682, 68693),List(),July/August)"
"TimEx(Interval(68687, 68693),List(),August)" <- missing

// One TimEx is inside another
"TimEx(Interval(70123, 70139),List(),October/November)"
"TimEx(Interval(70131, 70139),List(),November)" <- missing

// One TimEx is inside another
"TimEx(Interval(70267, 70283),List(),December/January)"
"TimEx(Interval(70276, 70283),List(),January)" <- missing

"TimEx(Interval(79233, 79241),List(),December)"
"text" : "October-December",
"startOffset" : 79225,
"endOffset" : 79241,

// One TimEx is inside another
"TimEx(Interval(80665, 80681),List(),October/November)"
"TimEx(Interval(80673, 80681),List(),November)" <- missing

"TimEx(Interval(106447, 106455),List(),December)"
"text" : "January-December",
"startOffset" : 106439,
"endOffset" : 106455,

"TimEx(Interval(107618, 107622),List(TimeStep(2016-01-01T00:00,2017-01-01T00:00)),2016)"
"text" : "2014-2016",
"startOffset" : 107613,
"endOffset" : 107622,

// One TimEx is inside another
"TimEx(Interval(119467, 119493),List(),February, June and October)"
"TimEx(Interval(119477, 119481),List(),June)" <- mising


"TimEx(Interval(128141, 128145),List(TimeStep(2015-01-01T00:00,2016-01-01T00:00)),2015)"
"text" : "2009-2015",
"startOffset" : 128136,
"endOffset" : 128145,

"TimEx(Interval(129210, 129214),List(),June)"
"text" : "February-June",
"startOffset" : 129201,
"endOffset" : 129214,

"TimEx(Interval(135050, 135054),List(TimeStep(2015-01-01T00:00,2016-01-01T00:00)),2015)"
"text" : "2010-2015",
"startOffset" : 135045,
"endOffset" : 135054,

"TimEx(Interval(135560, 135562),List(),me)"
"text" : "time",
"startOffset" : 135558,
"endOffset" : 135562,

// One TimEx is inside another
"TimEx(Interval(192799, 192815),List(),August/September)"
"TimEx(Interval(192806, 192815),List(),September)" <- missing
```

[10_FAO_a-i5505e.txt](https://github.com/clulab/eidos/files/3469342/10_FAO_a-i5505e.txt)
